### PR TITLE
d3-array: strict histogram proposition [BREAKING CHANGE]

### PR DESCRIPTION
**Pull request by denisname**
**Breaking Changes**

The `ThresholdArrayGenerator` interface has been **removed** and replaced by two disticts interfaces. For numbers thresholds use `ThresholdNumberArrayGenerator` and for dates thresholds use `ThresholdDateArrayGenerator`.

The `HistogramGenerator` interface is **deprecated**. Use `HistogramGeneratorDate` for `Date` values and `HistogramGeneratorNumber` for `Number` values.

All but one of `thresholds()` methods in `HistogramGenerator` have been **removed**. If you need them, update your code to use `HistogramGeneratorDate` and `HistogramGeneratorNumber`.

-----------------------------

Initial commit:

This pull request make histogram ready for strict null check (related: #14494). The changes are to be discussed... Strict function types and TS 2.3 are kept for another pull request.
Some backward incompatibilities are brought with this pull request. But, we can add some default generics...

Changes are made according to the following `d3.histogram` behaviours.

* The `histogram()` function accept an array containing `undefined`. Those will be filtered out.
* The `value` method may return `undefined`. This is useful to filter some values.
* The properties `Bin#x0` and `Bin#x1`may be undefined when histogram is computed on empty list or list with only `undefined`.
* Finally, for `domain`, we need a union with `[undefined, undefined]` to allow using `d3.extent` as domain.
Although this is useless when domain is `Date`. So we may want to divide `HistogramGenerator` in `HistogramGeneratorNumber` and `HistogramGeneratorDate` with different `domain` definition.
This also allows to remove `thresholds(count: ThresholdCountGenerator): this` from the _Date histogram_.

What are your thoughts ?

Related: #23611.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: see d3 api and source code
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by denisname** [comment]
@typescript-bot This PR needs to be discussed. Please don't merge.

**Comment by tomwanzek** [comment]
@denisname Sorry, I have been rather tied down recently, I won't get to it before second half of the week. Appreciate all your hard work though!

File: types/d3-array/index.d.ts
Comment (Ledragon) [inline types/d3-array/index.d.ts]: Could Value be used as the default generic type at definition level?

**Comment by Ledragon** [comment]
Is there a way a default could be used for the new generic parameter? As this would cause a breaking change
Seems ok to me otherwise

**Review by tomwanzek — COMMENTED** [review]
In principle, I like the refinement of the `HistogramGeneratorXYZ`, the main concern is of course the breaking nature of the change, i.e. removing `HistogramGenerator`.

I'll run a quick check through a couple of scenarios trying to retain a deprecated `HistogramGenerator<...>` as a union type (Noting that we are working with a max version of TS 2.3 right now)

I think, we might already want to switch on `strictFunctionTypes` to ensure nothing unforeseen happens with the refactoring.

In any case, when we're decided, we should definitely close to book on completing the JSDoc comments for this final bit of *d3-array*.

So will be back on that shortly, but wanted to give you the interim feedback.

File: types/d3-array/index.d.ts
Comment (denisname) [reply types/d3-array/index.d.ts]: This will not be backward compatibility for people working in strict function type. The exact equivalent with previous code is `= number>`. Should we care?

File: types/d3-array/index.d.ts
Comment (denisname) [reply types/d3-array/index.d.ts]: `Value` is the value accessor return type. If someone use a value accessor which always returns `number` and also may pass an empty list to `histogram(.)` at some point, we will force him to use `number | undefined` instead as `Value` generic.

Also note that all over in d3-array, we have functions returning `undefined` on empty list, and which therefore need a type assertion for most common use. I cannot tell you how many times I wrote: `d3.max(...)!` or `d3.min(...)!`. So I find this consistant.

Of course in general and for most users, it would be simpler as you propose...

**Comment by denisname** [comment]
> to retain a deprecated `HistogramGenerator<...>`.

I don't think there is a way to do that. There are alternatives however:

* Don't take in account the last point of my first comment ("Finally, for `domain`...") so we don't break the interface. Although less precise, type will still be better. And we can always do this split later.
* Use a type alias for the most common case (see code). It will break for `Date`but keep working for numbers.

```typescrip
/**
 * @deprecated Use HistogramGeneratorNumber<...> instead.
 * For Date values, use HistogramGeneratorDate<...>.
 */
export type HistogramGenerator<Datum, Value extends number> =
    HistogramGeneratorNumber<Datum, Value>;
```

> I think, we might already want to switch on `strictFunctionTypes` [...]

I have this ready. I will append a commit on this pull request, once we agreed on the `strictNullCheck`.

File: types/d3-array/index.d.ts
Comment (tomwanzek) [reply types/d3-array/index.d.ts]: On this one, I think it is fair game to change to `<Value extends number | undefined = number | undefined>` and add the default including `undefined`.

Adding generics without a default is definitely backwards breaking, see the necessary changes to *d3-contour* you already had to make. These would not be required.

W.r.t. backwards breaking under `strictFunctionTypes`:

While unfortunate that we are still working through the technical debt of all the validations for `strictNullChecks` and `strictFunctionTypes` (albeit at a much faster clip with your help), we always knew that the added strictness may be breaking folks who used `strictXYZ` already with the unvalidated definitions. That was also the clear reason not to update the `tsconfig` and portrait them as `strictXYZ` before having done the actual work.

File: types/d3-array/index.d.ts
Comment (tomwanzek) [reply types/d3-array/index.d.ts]: As you point out the scenario is slightly different from the one in **d3-geo**, i.e. the empty list argument you make w.r.t. `histogram(...)`, so let's leave it as is here.

I know, what you are saying about `!`-assertions in other uses of **d3-array**, I just prefer to minimize the need for it, whenever, there is a well-defined declarative approach.

More broadly, I find the assertions can easily get out of hand and can mask mistakes, in particular when re-factoring upstream code which then violates the assumptions that once justified the assertion. :smile:

**Comment by tomwanzek** [comment]
@denisname I might have a deprecation approach that might allow us to bridge the transition. Will test it out today and provide an updated here.

**Comment by tomwanzek** [comment]
@denisname I think the following may serve as a transition period solution:

(1) Let's keep all the suggestions you made for splitting the type-specific histogram generators in combination with the three `histogram<...>()` signatures in the current PR. This is really the future use.
(2) Add back in the original `HistogramGenerator<Datum, Value extends number | Date | undefined>` modified to work with `strictNullChecks` and `strictFunctionTypes`. Mark it as `@deprecated` with appropriate references to use `HistogramGeneratorNumber<...>` or `HistogramGeneratorDate<...>`
(3) Add a fourth *final* signature as a backwards compatibility fallback:
```ts
/**
 * @deprecated Do not use Value generic which mixes number and Date types. Use either number or Date 
 * (in combination with undefined, as applicable) to obtain a type-specific histogram generator.
 */
export function histogram<Datum, Value extends number | Date | undefined>(): HistogramGenerator<Datum, Value>;
```

For all intends and purposes, the fourth an final fallback will never be relevant in practice, as I doubt someone ever wrote `histogram<MyDatum, number | Date>()` That would have to be an attempt to implement a _very_ generic solution to a problem, where it is not known in advance whether `Date` or `number` is used, and even then they might have approached it differently.

It does leave some spurious code in for now, but should build a workable bridge to use your otherwise improved code.

If we ever come to a mutually agreeable and supported solution to #24999, even just using the `README.md`, then that would be the place to advise of the deprecations and a timeframe for removal.

**Comment by denisname** [comment]
I committed some code.

Add default generic for `ThresholdCountGenerator`, should I remove my changes from `d3-contour`?

Add the deprecated interface you proposed and a small test. But I had to remove the two `thresholds(count: ...): this;` from `HistogramGenerator<.>`. Otherwise the tests would fail...

**Comment by denisname** [comment]
@tomwanzek have you seen my last commit? Any recommendations ?

**Comment by armanio123** [comment]
@tomwanzek Got any chance to review the latest changes? Is it good to merge?

**Comment by denisname** [comment]
As previously [agreed](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25450#pullrequestreview-119669376), I added strict function type support.

* For statistics functions, to be able to write something like:
	
	```typescript
	d3.max([1, 2, 3], (d: number, i: number, arr: number[]) => arr[i]);
	```
	
	We must allow the array accessor to be of the same type as the given array. One way to do that is having three overloading, one for each array kind: `Array<.>`, `ReadonlyArray<.>` and `ArrayLike<.>`. Or let's just add a `K extends AnArray<T>` with `AnArray<T> = Array<T> | ReadonlyArray<T> | ArrayLike<T>`. Like this:
	
	```typescript
	export function max<T, U extends Numeric, K extends AnArray<T>>(array: K, accessor: (datum: T, index: number, array: K) => U | undefined | null): U | undefined;
	```
	
	Same modification as been done for every statistics functions. I added some tests for strict function type only for `d3.max`. Not sure if it useful to have it for every statistics functions. Also `$ExpectError` does not always work well with the build system, had to comment one out.

* For histogram, changed the `Array<.>` to `ArrayLike<.>` in `value(.)` as this was already the case with the [old histogram code](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/d3-array/index.d.ts#L331-L332).

* For histogram, also had to change `ThresholdCountGenerator` and `ThresholdArrayGenerator`. For the latter I could not do it without splitting the interface in two: `ThresholdDateArrayGenerator` and `ThresholdNumberArrayGenerator`. Which is analogue to the splitting of the `HistogramGenerator`.

I could not keep two of the thresholds functions in the deprecated interface. Having a complete backward compatible interface, with `strictNullChecks`, `strictFunctionTypes` and only TS 2.3 seems hard to do...

I updated `d3-contour`. This must be checked as I know mostly nothing about that library.

**Comment by paulvanbrenk** [comment]
🔔@gustavderdrache @borisyankov @tomwanzek @Ledragon since I don't have nearly enough context to say anything meaningful about this PR

**Comment by Ledragon** [comment]
@denisname do you feel confident enough to get this merged, even without @tomwanzek approval?

**Comment by denisname** [comment]
The type inference for accessors in statistics function was not working right. This is a minimal example:

```typescript
type AnArray<T> = T[] | ReadonlyArray<T> | ArrayLike<T>;
declare function min<T, K extends AnArray<T>>(array: K, accessor: (datum: T) => number): [number, number];

// inferred as min<number, number[]>
min([1, 2, 3], (d: number) => {
    const n: number = d; // OK
    return n;
});

// inferred as min<{}, number[]>, should be min<number, number[]>
min([1, 2, 3], (d) => {
    const n: number = d; // Error: Type '{}' is not assignable to type 'number'.
    return n;
});
```

I just changed the signature by using an index signature:

```typescript
function min<K extends AnArray<any>>(array: K, accessor: (datum: K[0]) => number): [number, number];
```

**Comment by denisname** [comment]
@Ledragon, I did some pretty extensive testing. So I'm confident. But having  @tomwanzek approval would be nice. A bug is always possible.

There is a known _backward incompatibility_ if someone use the `HistogramGenerator` interface and call the `threshold` function on it. But if `threshold` is called on a variable with inferred type, then there will be no problem. So few people should be affected. And it's easy to fix: use `HistogramGeneratorNumber` or `HistogramGeneratorDate` interface.

**Comment by tomwanzek** [comment]
Digesting the latest changes with `strictFunctionTypes`... ⏳

**Comment by tomwanzek** [comment]
First of, apologies for the delay in circling back to this. My first comment would be this: the considered changes to the _statistics_ functions are breaking in that they materially change the meaning (and even number of generics), so we should try to avoid it.

I think we should accept the implications of what the `strictFunctionTypes` mean at the root. I.e. if the "array" can be `AnArray<T>`, then we should align the accessor function to also accept `AnArray<T>`. Given that the accessor function should really just "access" existing data elements and not mutate them or the array, we are perfectly fine to require the following:
```ts
export function max<T>(array: AnArray<T>, accessor: (datum: T, index: number, array: AnArray<T>) => string | undefined | null): string | undefined;
export function max<T, U extends Numeric>(array: AnArray<T>, accessor: (datum: T, index: number, array: AnArray<T>) => U | undefined | null): U | undefined;
```
Yes, this means, that if anyone has explicitly defined accessor functions, as the tests do, and they switch on `strictFunctionTypes`, they would have to "fix" the type of the array element to be `AnArray<T>` as well.
However, any usage like:
```ts
numOrUndefined = d3Array.max(mixedObjectArray, (d, i, arr) => {
    return d.num;
});
```
or even just 
```ts
numOrUndefined = d3Array.max(mixedObjectArray, d => {
    return d.num;
});
```
will be totally unaffected.

File: types/d3-array/index.d.ts
Comment (tomwanzek) [reply types/d3-array/index.d.ts]: I guess we can even go back down to:
```ts
export function max<T>(array: ArrayLike<T>, accessor: (datum: T, index: number, array: ArrayLike<T>) => string | undefined | null): string | undefined;
export function max<T, U extends Numeric>(array: ArrayLike<T>, accessor: (datum: T, index: number, array: ArrayLike<T>) => U | undefined | null): U | undefined;
```
And simply fix the accessor functions in the test file to use `ArrayLike<MixedObject>` rather than `MixedObject[]`

File: types/d3-array/index.d.ts
Comment (tomwanzek) [reply types/d3-array/index.d.ts]: Are we changing this to `export function histogram(): HistogramGeneratorNumber<number, number | undefined>;` for the default.

**Review by tomwanzek — CHANGES_REQUESTED** [review]
Beyond my comments on the _statistics_ functions, I guess we are stuck with two breaking changes, even if split the histogram generators:

(1) Removing the `ThresholdArrayGenerator` type
(2) Removing several `threshold` related method overloads from the legacy `HistogramGenerator` interface.

Anyone who uses these two explicitly will break. Anyone, who just uses the return type of `histogram` should be fine.

Would love to already add a README with changelog for this change. It's probably the biggest breaking change we have ever introduced outside of action D3 breaking changes in accordance with semantic versioning.

**Comment by tomwanzek** [comment]
@andy-ms: related to #24999, we could really use a more expanded README to add a changelog section for this change. It's a major improvement to the definitions file, but it will be breaking and we have no way to wait for a "major" release of `d3-array` to introduce it.

So it would be great to have something to add a _previously_/_now use_ kind of guidance for updating the code.

@denisname At a minimum, we should update this PR (title and initial merge comment) to call out a BREAKING CHANGE and add a summary of the breaking change with migration description at the top of the opening comment to this PR.

**Comment by denisname** [comment]
> Are we changing this to export function histogram(): HistogramGeneratorNumber<number, number | undefined>; for the default.

As you want @tomwanzek, but it may create more backward incompatibilities.

**Comment by RyanCavanaugh** [comment]
@denisname @tomwanzek @Ledragon should I merge this?

**Comment by denisname** [comment]
@tomwanzek I added a summary of the breaking in the top comment.

We only still need to decide for the default generic:
```typescript
export function histogram(): HistogramGeneratorNumber<number, number>;
```
or
```typescript
export function histogram(): HistogramGeneratorNumber<number, number | undefined>
```

Once we have decided, we will be able to close this pull request.

**Comment by Ledragon** [comment]
@denisname @tomwanzek where are we standing on this?

**Comment by denisname** [comment]
@RyanCavanaugh can we merge this PR.
There has been no reaction from @tomwanzek for more than three months.
We mostly found an agreement anyway. There is only one small [detail](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25450#issuecomment-409190062) left.
But we can always relax the typing later if necessary.

**Pull request merged**
